### ISSUE TITLE
Floating-point inconsistency

### ISSUE THREAD

**Issue by Baltoli** [runtimeverification/k#2312]
The same version of MPFR seems to produce different results on ARM and X86 when performing exponentiation and logarithms; investigate why this is the case.

**Comment by Baltoli** [comment]
Failing tests are unit tests for unary operations (`e^x`, `log(x)`)

**Comment by Baltoli** [comment]
This isn't in fact an issue with MPFR (which produces consistent results across both platforms). The problem is with the Java standard library, which produces different results on x86 and ARM.

For example:
```java
System.out.println(Math.exp(1.0));
```
will print `2.718281828459045` on x86, but `2.718281828459045[5]` on ARM (note the emphasised extra 5 at the end).

Interestingly, the ARM version is in line with the JVM's `StrictMath` implementation. The following program demonstrates this:
```java
System.out.println("Strict : " + StrictMath.exp(1.0));
System.out.println("Default: " + Math.exp(1.0));
```

```console
arm $ ./run
Strict : 2.7182818284590455
Default: 2.7182818284590455
x86 $ ./run
Strict : 2.7182818284590455
Default: 2.718281828459045
```

The possible solutions here that I can see are:
* Update the `BigFloat` implementation to be consistent with `StrictMath` somehow, and update the reference functions in the test suite.
  * This might break other tests if they aren't compatible with `StrictMath`.
* Decrease the precision of the reference checks on these tests (i.e. to some number of ULPs).

**Comment by radumereuta** [comment]
A few ideas:
1. Did you search the web for other people having the same issue?
2. What versions of java are you using? If they are different/old, what happens if you update?

**Comment by Baltoli** [comment]
1. It's reasonably [well-documented](https://stackoverflow.com/questions/32977034/in-java-is-strictmath-still-preferred-to-math-for-consistency) that `Math` can perform platform-specific optimisations in the JVM. I'm not well enough informed to say why the faster / less precise version matches the MPFR behaviour.
2. I tested OpenJDK 11 (supplied by `apt`) and 17 (downloaded binary) on my x86 Ubuntu machine and got the same results.
    1. Version 17 on ARM macOS behaves differently to 17 on x86 Ubuntu.
    2. Version 17 on x86 Ubuntu behaves the same as 17 on x86 macOS.
    3. This seems to suggest it's a platform / architecture issue, rather than a Java one.

**Comment by Baltoli** [comment]
Interestingly, this doesn't manifest in C programs:

```c
#include <math.h>
#include <stdio.h>

int main(void) {
  printf("%.17g\n", exp(1.0));
}
```
will print `2.7182818284590451` on all the platforms I tested, which suggests the non-strict JVM implementation on x86 is (perhaps) calling into C.

This presents an option to fix the tests - rather than using Java's `Math::exp`, we could make a JNI call into `exp(double)`.

**Comment by Baltoli** [comment]
Closed by #2319
